Revise sandbox to allow necessary sysctl calls

brentfulgham opened the pull request:
#### b64091ecce1bc486b4621fcbcdfe4c047d9093d8
<pre>
Revise sandbox to allow necessary sysctl calls
<a href="https://rdar.apple.com/107428604">rdar://107428604</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=271899">https://bugs.webkit.org/show_bug.cgi?id=271899</a>

Reviewed by Per Arne Vollan.

Telemetry has revealed a sysctl-read call that is needed in normal use:

    kern.willshutdown

We can also silence a few sysctl-read calls that are only needed for use
cases where WebKit does not run:

    sysctl.proc_translated
    vm.page_wire_count

We should add them to the sandbox to avoid spurious logging.

* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Source/WebKit/webpushd/mac/com.apple.WebKit.webpushd.mac.sb.in:
* Source/WebKit/webpushd/mac/com.apple.WebKit.webpushd.relocatable.mac.sb.in:

Canonical link: <a href="https://commits.webkit.org/276953@main">https://commits.webkit.org/276953@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8aff16e35d2ecde74bebb6427d35b42dd3161e4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48838 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42208 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29626 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22741 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37724 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46745 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22365 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39817 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18930 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19760 "Found unexpected failure with change (failure)") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4211 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42463 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41264 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50634 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21163 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17656 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44898 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22463 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43806 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10241 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22822 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22157 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
```diff
@@ -222,6 +222,8 @@
         "kern.ostype"
         "kern.version"
+        "kern.willshutdown"
+        "sysctl.proc_translated"
```
Are these in use on iOS too? It looks like reports are on macOS so far?

brentfulgham replied:
`kern.willshutdown` is definitely used on iOS (and is in other system sandboxes).

You are right that `sysctl.proc_translated` is very likely NOT. I'll fix that.

brentfulgham replied:
Actually -- I take that back. There is lots of code on iOS that checks it (even though it would likely only ever return `false`.) I think returning an honest `false` without failures (and paying the cost of error logging) is more desirable than relying on the sys call failure to produce expected behavior.

So I think we should just leave both of these enabled in the sandbox.

pvollan replied:
Do we think they are in use in the WebKit processes? Perhaps we should add sysctl telemetry to determine if they are used before potentially adding to the sandbox?

brentfulgham replied:
I have to admit that I do not find hits for either of them in iOS telemetry, so I will remove them from this patch.

pvollan replied:
Thank you for revising the patch! I wonder if these are strictly required, and we could instead silence them, at least in the case for the WebKit processes? If we go down that route, we could add a common silencing rule in `WebKit/Shared/Sandbox/macOS/common.sb`.

pvollan commented on Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
```diff
@@ -221,7 +221,9 @@
         "kern.osproductversion"
         "kern.osrelease"
         "kern.ostype"
+        "kern.osversion"
```
Is this also required?

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -453,7 +453,7 @@
         "machdep.cpu.brand_string"
         "security.mac.sandbox.sentinel"
-        "sysctl.name2oid"
+        "sysctl.name2oid"        
```
This change appears to be unintentional.

pvollan reviewed (approved):
Very nice, LGTM, thank you for revising the patch! R=me.

brentfulgham committed:
Revise sandbox to allow necessary sysctl calls
rdar://107428604
https://bugs.webkit.org/show_bug.cgi?id=271899

Reviewed by Per Arne Vollan.

Telemetry has revealed a sysctl-read call that is needed in normal use:

    kern.willshutdown

We can also silence a few sysctl-read calls that are only needed for use
cases where WebKit does not run:

    sysctl.proc_translated
    vm.page_wire_count

We should add them to the sandbox to avoid spurious logging.

* Source/WebKit/GPUProcess/mac/com.apple.WebKit.GPUProcess.sb.in:
* Source/WebKit/NetworkProcess/mac/com.apple.WebKit.NetworkProcess.sb.in:
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Source/WebKit/webpushd/mac/com.apple.WebKit.webpushd.mac.sb.in:
* Source/WebKit/webpushd/mac/com.apple.WebKit.webpushd.relocatable.mac.sb.in:

Canonical link: https://commits.webkit.org/276953@main
